pyms/flask/services/service_discovery.py

```diff
@@ -0,0 +1,82 @@
+import json
+import logging
+import uuid
+
+import requests
+from requests.exceptions import RequestException
+from pyms.constants import LOGGER_NAME
+from pyms.exceptions import ServiceDiscoveryConnectionException
+from pyms.flask.services.driver import DriverService
+
+logger = logging.getLogger(LOGGER_NAME)
+
+CONSUL_SERVICE_DISCOVERY = "consul"
+
+DEFAULT_SERVICE_DISCOVERY = CONSUL_SERVICE_DISCOVERY
+
+
+class ServiceDiscoveryBase:
+    def register_service(self, id_app, host, port, healtcheck_url, app_name):
+        pass
+
+
+class ServiceDiscoveryConsul(ServiceDiscoveryBase):
+    def register_service(self, id_app, host, port, healtcheck_url, app_name):
+        headers = {
+            'Content-Type': 'application/json; charset=utf-8'
+        }
+        data = {
+            "id": app_name + "-" + id_app,
+            "name": app_name,
+            "port": port,
+            "check": {
+                "name": "ping check",
+                "http": healtcheck_url,
+                "interval": "30s",
+                "status": "passing"
+            }
+        }
+        error = False
+        msg_error = "Failed to establish a new connection"
+        try:
+            response = requests.put("{host}/v1/agent/service/register".format(host=host), data=json.dumps(data),
+                                    headers=headers)
+            if response.status_code != 200:
+                msg_error = response.content
+                error = True
+        except RequestException:
+            error = True
+
+        if error:
+            raise ServiceDiscoveryConnectionException("Host %s raise an error: %s" % (host, msg_error))
+
+
+class Service(DriverService):
+    id_app = str(uuid.uuid1())
+    config_resource = "service_discovery"
+    default_values = {
+        "client": DEFAULT_SERVICE_DISCOVERY,
+        "host": "http://localhost:8500",
+        "port": 5000,
+        "healtcheck_url": "http://127.0.0.1.nip.io:5000/healthcheck",
+    }
+
+    def init_action(self, microservice_instance):
+        self.client.register_service(
```

Comment:
I'd argue that registration is not an obligatory step of the service discovery. While consul expects services to register themselves it may be done from consul's site. In Docker or Kubernetes env vars or DNS may be used - those don't need client-side registration action (I'm planning to implement those two to check how would it work in Docker Compose).
So the question is: should client like that implement empty registration method or instead there should be static variable that should decide if call is necessary? What do you think?

Reply:
Do you mean that could be an option in the configuration to autoregister instead to register in all cases? in example:

```yaml
    service_discovery:
      client: "consul"
      host: "http://localhost:8500"
      autoregister: true
```

Reply:
Yeah. Maybe just key ‘register’ and registration params as values? No key = no registration.